### PR TITLE
gh-180 Generate valid Signature for BiFunction's wrapped by ApplyBiFu…

### PR DESCRIPTION
**Pull request by m29827**
…nction

### CODE REVIEW

File: core/src/main/java/uk/gov/gchq/koryphe/signature/Signature.java [hunk 1]
```diff
@@ -112,25 +116,93 @@ public static Signature getOutputSignature(final Function function) {
      * @return Signature of the type variable.
      */
     private static Signature createSignatureFromTypeVariable(final Object input, final Class functionClass, final int typeVariableIndex, final boolean isInput) {
+
```
Comment (d47853) [inline core/src/main/java/uk/gov/gchq/koryphe/signature/Signature.java:119]:
```suggestion
```

File: core/src/main/java/uk/gov/gchq/koryphe/signature/Signature.java [hunk 2]
```diff
@@ -112,25 +116,93 @@ public static Signature getOutputSignature(final Function function) {
...
+            typeArgs.putAll(inputToWrappedBiFunctionTypeArgs);
+        }
+
         Type type = typeArgs.containsKey(tv) ? typeArgs.get(tv) : Object.class;
+
         return createSignature(input, type, typeArgs, isInput);
     }
 
+    private static Map<TypeVariable<?>, Type> createInputToWrappedBiFunctionTypeArgMapping(final Object input, final Map<TypeVariable<?>, Type> typeArgs, final Map<TypeVariable<?>, Type> wrappedBiFunctionTypeArgs) {
+
```
Comment (d47853) [inline core/src/main/java/uk/gov/gchq/koryphe/signature/Signature.java:144]:
```suggestion
```

File: core/src/main/java/uk/gov/gchq/koryphe/signature/Signature.java [hunk 3]
```diff
@@ -112,25 +116,93 @@ public static Signature getOutputSignature(final Function function) {
...
     private static Signature createSignature(final Object input, final Type type, final Map<TypeVariable<?>, Type> typeArgs, final boolean isInput) {
         Class clazz = null;
-        if (input.getClass().getTypeParameters().length > 0) {
+
+        if (KorypheFunctionN.class.isAssignableFrom(input.getClass())) {
+
+            clazz = getTypeClass(type, typeArgs);
+
+        } else if (input.getClass().getTypeParameters().length > 0) {
+
```
Comment (d47853) [inline core/src/main/java/uk/gov/gchq/koryphe/signature/Signature.java:199]:
```suggestion
```

File: core/src/test/java/uk/gov/gchq/koryphe/signature/SignatureTest.java [hunk 4]
```diff
@@ -271,4 +274,38 @@ public void ShouldCheckCollectionConcatInputAndOutput() {
         // Then
         assertTrue(output.assignable(Collection.class).isValid());
     }
+
+    @Test
+    public void shouldCheckApplyBiFunctionTypes() {
+
```
Comment (d47853) [inline core/src/test/java/uk/gov/gchq/koryphe/signature/SignatureTest.java:280]:
```suggestion
```

File: core/src/test/java/uk/gov/gchq/koryphe/signature/SignatureTest.java [hunk 5]
```diff
@@ -271,4 +274,38 @@ public void ShouldCheckCollectionConcatInputAndOutput() {
...
+        assertTrue(signature.assignable(Long.class, Long.class).isValid());
+
+        assertFalse(signature.assignable(Integer.class).isValid());
+        assertFalse(signature.assignable(Integer.class, Integer.class, Integer.class).isValid());
+        assertFalse(signature.assignable(String.class, String.class).isValid());
+    }
+
+    @Test
+    public void shouldCheckApplyBiFunctionTypesForInlineBiFunction() {
+
```
Comment (d47853) [inline core/src/test/java/uk/gov/gchq/koryphe/signature/SignatureTest.java:294]:
```suggestion
```

File: core/src/main/java/uk/gov/gchq/koryphe/signature/Signature.java [hunk 6]
```diff
@@ -112,25 +116,93 @@ public static Signature getOutputSignature(final Function function) {
...
+        for (final Map.Entry<TypeVariable<?>, Type> entry : wrappedBiFunctionTypeMap.entrySet()) {
+            if (BiFunction.class.equals(getClassFrom(entry.getKey().getGenericDeclaration()))) {
+                mapping.put(entry.getKey(), entry.getValue());
+            }
+        }
+        return mapping;
+    }
+
+    private static Class<?> getClassFrom(final GenericDeclaration genericDeclaration) {
+
```
[the remark this answers is not in the document]
Comment (m29827) [reply core/src/main/java/uk/gov/gchq/koryphe/signature/Signature.java:184]: Ok, done

**Pull request merged**